src/ir/gc-type-utils.h

```diff
@@ -70,6 +70,19 @@ inline EvaluationResult evaluateCastCheck(Type refType, Type castType) {
 
   auto refHeapType = refType.getHeapType();
   auto castHeapType = castType.getHeapType();
+
+  if (castHeapType.isBottom()) {
+    // We are casting to a null type, which means we are checking for a null.
+    // Both the reference must be nullable (so a null can appear) and the cast
+    // type as well (or else the cast type is an impossible type, a non-nullable
+    // null).
+    if (refType.isNullable() && castType.isNullable()) {
+      return SuccessOnlyIfNull;
+    } else {
+      return Failure;
+    }
+  }
+
   auto refIsHeapSubType = HeapType::isSubType(refHeapType, castHeapType);
   auto castIsHeapSubType = HeapType::isSubType(castHeapType, refHeapType);
   bool heapTypesCompatible = refIsHeapSubType || castIsHeapSubType;
```

Comment:
This smaller diff would accomplish the same thing.
```suggestion
  // The dynamic heap type of the reference might be a subtype of the cast type only if the types are related and the cast type is instantiable.
  bool heapTypesCompatible = (refIsHeapSubType || castIsHeapSubType) && !castHeapType.isBottom();
```

Reply:
What does "instantiable" mean? A null is a real value, so I guess you mean "not created by struct.new" or such?

But even so I'm not sure how to read the English here. Why can't the cast type be a bottom type for the types to be compatible? A null input value would be compatible with a null cast type, wouldn't it?

Reply:
I do see how the logic is repeated, though... but I'm not sure how to explain a merge of the code...

Reply:
I pushed a commit that unifies the code, but now the comment is twice as long :thinking:

Reply:
It’s not the reference type but rather the bottom heap type that is not instantiable, i.e. not able to be instantiated.

“compatible” here means it’s possible for the dynamic heap type of the referenced object to be a subtype of the cast target heap. If the referenced heap type is bottom, the referenced object cannot exist so the heap types cannot be compatible.

Reply:
I guess you're special-casing null in the last paragraph? In my mind a null value **is** compatible with a null cast target. Why is that wrong?

That is, no instantiated object exists (as you said) but a null is still a real value, and it has a heap type, and that heap type is compatible with the cast type's heap type. So the heap types are compatible.

Reply:
FWIW I'm fine with the way you've written it as well, where special casing bottom is done separately from the computation of "compatibility."